src/Processors/QueryPlan/Optimizations/optimizeJoinByLayers.cpp

```diff
@@ -0,0 +1,412 @@
+
+    for (const auto & output : lhs_dag.getOutputs())
+        lhs_outputs.emplace(output->result_name, output);
+    for (const auto & output : rhs_dag.getOutputs())
+        rhs_outputs.emplace(output->result_name, output);
+
+    const auto & lhs_pk_dag = lhs_pk.expression->getActionsDAG();
+    const auto & lhs_pk_colum_names = lhs_pk.column_names;
+    auto lhs_matches = matchTrees(lhs_pk_dag.getOutputs(), lhs_dag, false);
+    const auto & rhs_pk_dag = rhs_pk.expression->getActionsDAG();
+    const auto & rhs_pk_colum_names = rhs_pk.column_names;
+    auto rhs_matches = matchTrees(rhs_pk_dag.getOutputs(), rhs_dag, false);
+
+    JoinStep::PrimaryKeySharding sharding;
+
+    for (size_t pos = 0; pos < lhs_pk_colum_names.size() && pos < rhs_pk_colum_names.size(); ++pos)
+    {
+        // std::cerr << "Checking pos " << pos << std::endl;
+
+        const auto * lhs_pk_output = lhs_pk_dag.tryFindInOutputs(lhs_pk_colum_names[pos]);
+        const auto * rhs_pk_output = rhs_pk_dag.tryFindInOutputs(rhs_pk_colum_names[pos]);
+
+        if (!lhs_pk_output || !rhs_pk_output)
+            break;
+
+        size_t keys_size = clause.key_names_left.size();
+
+        for (size_t i = 0; i < keys_size && sharding.size() <= pos; ++i)
+        {
+            const auto & left_name = clause.key_names_left[i];
+            const auto & right_name = clause.key_names_right[i];
+
+            // std::cerr << left_name << ' ' << right_name << std::endl;
+
+            auto it = lhs_outputs.find(left_name);
+            auto jt = rhs_outputs.find(right_name);
+            if (it == lhs_outputs.end() || jt == rhs_outputs.end())
+                continue;
+
+            auto lhs_match = lhs_matches.find(it->second);
+            auto rhs_match = rhs_matches.find(jt->second);
+            if (lhs_match == lhs_matches.end() || rhs_match == rhs_matches.end())
+                continue;
+
+            if (lhs_match->second.monotonicity || rhs_match->second.monotonicity)
+                continue;
+
+            if (lhs_match->second.node == lhs_pk_output && rhs_match->second.node == rhs_pk_output)
+            {
+                // std::cerr << lhs_pk_dag.dumpDAG() << std::endl;
+                // std::cerr << rhs_pk_dag.dumpDAG() << std::endl;
+                // std::cerr << "==== match\n";
+                sharding.emplace_back(lhs_pk_colum_names[pos], rhs_pk_colum_names[pos]);
+            }
+        }
+
+        if (sharding.size() <= pos)
+            break;
+    }
+
+    return sharding;
+}
+
+struct JoinsAndSourcesWithCommonPrimaryKeyPrefix
+{
+    struct JoinAndSharding
+    {
+        JoinStep * join;
+        JoinStep::PrimaryKeySharding sharding;
+    };
+
+    std::list<JoinAndSharding> joins;
+    std::list<JoinStep *> joins_to_keep_in_order;
+    std::list<ReadFromMergeTree *> sources;
+    std::list<SortingStep *> sorting_steps;
+    size_t common_prefix = std::numeric_limits<size_t>::max();
```

Comment:
It's not really easy to understand in which order do we expect have steps in these lists and how they correspond to each other, so comment may be helpful.